ui/index.d.ts

```diff
@@ -360,6 +362,128 @@ export declare abstract class AzureWizardStep<T> {
     public abstract execute(wizardContext: T, outputChannel: OutputChannel): Promise<T>;
 }
 
+export interface ISubscriptionWizardContext {
+    credentials: ServiceClientCredentials;
+    subscription: Subscription;
+}
+
+export interface ILocationWizardContext extends ISubscriptionWizardContext {
+    /**
+     * You may specify the defaultLocationName if you don't want the `LocationStep` to prompt for a location
+     * For example, if the user selects an existing resource, you might want to use that location as the default for the wizard's other resources
+     */
+    defaultLocationName?: string;
+
+    /**
+     * The location to use for new resources
+     * This value will be defined after `LocationStep.prompt` occurs.
+     */
+    location?: Location;
+}
+
+export declare class LocationStep<T extends ILocationWizardContext> extends AzureWizardStep<T> {
+    public prompt(wizardContext: T, ui: IAzureUserInput): Promise<T>;
+    public execute(wizardContext: T, outputChannel: OutputChannel): Promise<T>;
+}
+
+export interface IAzureNamingRules {
+    minLength: number;
+    maxLength: number;
+
+    /**
+     * A RegExp specifying the invalid characters.
+     * For example, /[^a-z0-9]/ would specify that only lowercase, alphanumeric characters are allowed.
+     */
+    invalidCharsRegExp: RegExp;
+
+    /**
+     * Specify this if only lowercase letters are allowed
+     * This is a separate property than `invalidCharsRegExp` because the behavior can be different.
+     * For example, when generating a relatedName, we can convert uppercase letters to lowercase instead of just removing them.
+     */
+    lowercaseOnly?: boolean;
+}
+
+export interface IRelatedNameWizardContext {
+    /**
+     * A task that evaluates to the related name that should be used as the default for other new resources or undefined if a unique name could not be found
+     * The task will be defined after `AzureNameStep.prompt` occurs.
+     */
+    relatedNameTask?: Promise<string | undefined>;
+}
+
+/**
+ * A generic class for a step that specifies the name of a new resource, used to generate a related name for other new resources.
+ * You must implement `isRelatedNameAvailable` and assign `wizardContext.relatedNameTask` to the result of `generateRelatedName`
+ */
+export declare abstract class AzureNameStep<T extends IRelatedNameWizardContext> extends AzureWizardStep<T> {
+    /**
+     * This method will by called by `generateRelatedName` when trying to find a unique suffix for the related name
+     * @param wizardContext The context of the wizard.
+     * @param name The name that will be checked.
+     */
+    protected abstract isRelatedNameAvailable(wizardContext: T, name: string): Promise<boolean>;
```

Comment:
I'm hesitant to create our own isRelatedNameAvailable functions.  The WebSiteManagementClient has its own checkNameAvailability function that should handle checking ResouceNameAvailability.  It seems like only the storage step is leveraging this.

The portal isn't super transparent about their naming restrictions so we may unintentionally miss/block resource names by creating our own rules.  Or if they alter their naming rules without alerting us.

Reply:
I see your point and I think we should make this decision on a case-by-case basis. I agree that ideally we can use `checkNameAvailability`, but frankly its pretty terrible sometimes. For example this is what the WebSiteManagementClient.checkNameAvailability message looks like:
![screen shot 2018-03-22 at 11 45 44 am](https://user-images.githubusercontent.com/11282622/37791580-c7a3aaac-2dc6-11e8-9219-4d69bf89371d.png)

The naming rules actually don't change that often and there's usually documentation outside of the portal. For example:
https://docs.microsoft.com/en-us/azure/architecture/best-practices/naming-conventions#storage

Reply:
Fair enough.  We could always intercept the error message and output actual human readable errors too, but not have to write/enforce our own rules.

But anyway, I don't consider this blocking as it's not something that would be hard to change later.

Reply:
We can't intercept their messages because that would assume English.  This is something unfortunately they've never done but should, but at least they document them.